Libplanet/Blocks/Block.cs

```diff
@@ -305,32 +314,50 @@ public int BytesLength
-                    nonce: Nonce.ToByteArray().ToImmutableArray(),
-                    miner: Miner?.ToByteArray().ToImmutableArray() ?? ImmutableArray<byte>.Empty,
-                    difficulty: Difficulty,
-                    totalDifficulty: TotalDifficulty,
-                    previousHash: previousHashAsArray,
-                    txHash: TxHash?.ToByteArray().ToImmutableArray() ?? ImmutableArray<byte>.Empty,
-                    hash: Hash.ToByteArray().ToImmutableArray(),
-                    preEvaluationHash: PreEvaluationHash.ToByteArray().ToImmutableArray(),
-                    stateRootHash: stateRootHashAsArray
-                );
+                return _header
+                    ?? (_preEvaluationHash is null
+                        ? (BlockHeader)(_header = new BlockHeader(
+                            protocolVersion: ProtocolVersion,
+                            index: Index,
+                            timestamp: Timestamp.ToString(
+                                BlockHeader.TimestampFormat,
+                                CultureInfo.InvariantCulture),
+                            nonce: Nonce.ToByteArray().ToImmutableArray(),
+                            miner: Miner?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty,
+                            difficulty: Difficulty,
+                            totalDifficulty: TotalDifficulty,
+                            previousHash: PreviousHash?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty,
+                            txHash: TxHash?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty))
+                        : (BlockHeader)(_header = new BlockHeader(
+                            protocolVersion: ProtocolVersion,
+                            index: Index,
+                            timestamp: Timestamp.ToString(
+                                BlockHeader.TimestampFormat,
+                                CultureInfo.InvariantCulture),
+                            nonce: Nonce.ToByteArray().ToImmutableArray(),
+                            miner: Miner?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty,
+                            difficulty: Difficulty,
+                            totalDifficulty: TotalDifficulty,
+                            previousHash: PreviousHash?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty,
+                            txHash: TxHash?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty,
+                            preEvaluationHash: PreEvaluationHash.ToByteArray().ToImmutableArray(),
+                            stateRootHash: StateRootHash?.ToByteArray().ToImmutableArray()
+                                ?? ImmutableArray<byte>.Empty)));
```

Comment:
```suggestion
                    ?? (BlockHeader)(_header = new BlockHeader(
                            protocolVersion: ProtocolVersion,
                            index: Index,
                            timestamp: Timestamp.ToString(
                                BlockHeader.TimestampFormat,
                                CultureInfo.InvariantCulture),
                            nonce: Nonce.ToByteArray().ToImmutableArray(),
                            miner: Miner?.ToByteArray().ToImmutableArray()
                                ?? ImmutableArray<byte>.Empty,
                            difficulty: Difficulty,
                            totalDifficulty: TotalDifficulty,
                            previousHash: _preEvaluationHash?.ToByteArray().ToImmutableArray()
                                ?? ImmutableArray<byte>.Empty,
                            txHash: TxHash?.ToByteArray().ToImmutableArray()
                                ?? ImmutableArray<byte>.Empty,
                            preEvaluationHash: PreEvaluationHash.ToByteArray().ToImmutableArray(),
                            stateRootHash: StateRootHash?.ToByteArray().ToImmutableArray()
                                ?? ImmutableArray<byte>.Empty)));
```

Reply:
This would not work. It is meant to use different constructors for `BlockHeader` using signature depending on whether `_preEvaluationHash` has been set or not. I know this part of the code seem problematic, but ultimately, constructors (and even classes) will be separated for half built `Block<T>` without `StateRootHash` and  fully built `Block<T>` with `StateRootHash`.

Reply:
In any case, due to rebasing and substantial amount of rework, this has become irrelevant. 😶